### PR TITLE
Computing resources branch

### PR DESCRIPTION
**Pull request by GernotMaier**
This pull request includes two changes:

1. Minor change to command line parameters to be more consistent with other applications. Main change is to use no `-p` or `--productionconfig` instead of `-c`/`--config` for the yaml file describing the parameters for the production. This hopefully helps to avoid confusing it with the `--configFile` option for generatic gammasim-tools configurations.

2. Addition of feature to provide an estimate of the run time / event for the production.py application.

This uses the `SECONDS` variable from unix bash (wallclock time) written into the bash submission scripts for corsika and sim_telarray.

A typical report is obtained after running a small local production e.g., with
`python applications/production.py -t simulate -p ../prodConfigTest.yml --primary gamma`

Then do:

```
python applications/production.py -t resources -p ../prodConfigTest.yml --primary gamma --showers_only

Computing Resources Report - Showers
#events/run = 50.00
Runtime/run [sec] = 15.50
Runtime/1000 events [sec] = 310.00
```
and
```
python applications/production.py -t resources -p ../prodConfigTest.yml --primary gamma --array_only

Computing Resources Report - Array Simulations
Runtime/run [sec] = 0.70
```

This is a first step for resource estimates and requires a bit more work in future (but not high priority now). There is also some inconsistency in the log file naming (log-out for corsika, sub-out for sim_telarray), which we want to address in future. Suggest to merge it now to make sure that the pull request does not get very big (and maybe look at higher priority issues now)

### CODE REVIEW

**Comment by GernotMaier** [comment]
@orelgueta - let's discuss before adding more time into this PR if we want to integrate it.

To goal is here to add functionality to with a flag `resources` to print overall run time and runtime per event to the screen. It relies on output from the qsub scripts (therefore are dependent on the batch system), but all my log files required to prepare this output are of zero file length.

Question: do we actually need this functionality, as we probably rely in future more on the CTA-Computing-Resources project or something similar. My tendency would be to discard the changes provided with this branch.

**Comment by orelgueta** [comment]
I am not sure. I can definitely imagine that we will want/need this functionality in the near and far future. I think the fact it relies on qsub is less than ideal, but the I think having this functionality is good. 

> Question: do we actually need this functionality, as we probably rely in future more on the CTA-Computing-Resources project or something similar. My tendency would be to discard the changes provided with this branch.

This sentence is a bit strange since you mention we will rely on this kind of calculation more in the future, and yet you suggest to discard these changes. Why is that?

**Comment by GernotMaier** [comment]
> I am not sure. I can definitely imagine that we will want/need this functionality in the near and far future. I think the fact it relies on qsub is less than ideal, but the I think having this functionality is good.
> 
> > Question: do we actually need this functionality, as we probably rely in future more on the CTA-Computing-Resources project or something similar. My tendency would be to discard the changes provided with this branch.
> 
> This sentence is a bit strange since you mention we will rely on this kind of calculation more in the future, and yet you suggest to discard these changes. Why is that?

I was referring to Rauls' https://github.com/RaulRPrado/CTA-Computing-Resources repository. Didn't you use that recently to estimate the required resources? Or was this something else?

**Comment by orelgueta** [comment]
Oh, I didn't understand that. This is not the same functionality tough, at least not the way I used the [CTA-Computing-Resources](https://github.com/RaulRPrado/CTA-Computing-Resources) repo. There I used it to estimate the necessary number of events to reach a certain level of uncertainty, while this branch allows us to calculate the computing resources per event which can be useful to estimate the cost of Prod6 for example.

File: simtools/simtel/simtel_runner_array.py [hunk 1]
```diff
@@ -168,6 +182,30 @@ def getOutputFile(self, run):
         )
         return self._simtelDataDir.joinpath(fileName)
 
+    def hasSubLogFile(self, run, mode='out'):
+        runSubFile = self.getSubLogFile(run=run, mode=mode)
+        return Path(runSubFile).is_file()
+
+    def getResources(self, run):
```
Comment (orelgueta) [inline simtools/simtel/simtel_runner_array.py:189]: This method looks like it's the same as the one before where I made a comment on the efficiency of reading, except in this case we are reading in reversed order which I assume would be much more efficient. Perhaps reading in reverse order is faster than my suggestion above.

Either way, can these two functions be made into one general one instead?

Also, docstring =)

**Review by orelgueta — COMMENTED** [review]
Most of my comments are about the docstring.

However, there are a few which are of more substance and discuss the duplication of functions and some other implementation suggestions. Nothing is critical though, so if you wish to merge and work on these things later, that's fine. We should open issues for them though in that case so that we don't forget.

File: simtools/array_simulator.py [hunk 2]
```diff
@@ -245,9 +241,20 @@ def submit(self, inputFileList, submitCommand=None, extraCommands=None, test=Fal
             runScript = self._simtelRunner.getRunScript(
                 run=run, inputFile=file, extraCommands=extraCommands
             )
-            self._logger.info("Run {} - Submitting script {}".format(run, runScript))
+            thisSubCmd = copy(subCmd)
+
+            # Checking for log files in sub command and replacing them
+            if 'log_out' in subCmd:
+                logOutFile = self._simtelRunner.getSubLogFile(run=run, mode='out')
+                thisSubCmd = thisSubCmd.replace('log_out', str(logOutFile))
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/array_simulator.py:249]: Ignore this - the submit part moves into a new classes with PR #308.

File: simtools/corsika/corsika_runner.py [hunk 3]
```diff
@@ -310,6 +316,38 @@ def _getExtraCommands(extra):
...
+        runNumber = self._validateRunNumber(runNumber)
+        runSubFile = self.getSubLogFile(runNumber=runNumber, mode=mode)
+        return Path(runSubFile).is_file()
+
+    def getResources(self, runNumber=None):
+        runNumber = self._validateRunNumber(runNumber)
+        subLogFile = self.getSubLogFile(runNumber=runNumber, mode='out')
+
+        runtime = None
+        with open(subLogFile, 'r') as file:
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/corsika/corsika_runner.py:334]: Is it necessary? These type of files have typically 1-2 lines only - efficiency is not an issue.

File: simtools/corsika/corsika_runner.py [hunk 4]
```diff
@@ -310,6 +316,38 @@ def _getExtraCommands(extra):
...
+                    runtime = int(line.split()[1])
+                    break
+
+        if runtime is None:
+            self._logger.debug('RUNTIME was not found in run log file')
+
+        # Calculating number of events
+        nEvents = int(
+            self.corsikaConfig.getUserParameter('NSHOW')
+            * self.corsikaConfig.getUserParameter('CSCAT')[0]
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/corsika/corsika_runner.py:346]: I think Raul wanted to implemented as well how much disk space is used - for that the reuse factor would be important. In the cause now, where only run time per event is printed not. I removed it.

File: simtools/shower_simulator.py [hunk 5]
```diff
@@ -227,12 +227,48 @@ def submit(
...
             )
-            self._logger.info("Run {} - Submitting script {}".format(run, runScript))
 
-            shellCommand = subCmd + " " + str(runScript)
+            thisSubCmd = copy(subCmd)
+
+            # Checking for log files in sub command and replacing them
+            if 'log_out' in subCmd:
+                logOutFile = self._corsikaRunner.getSubLogFile(runNumber=run, mode='out')
+                thisSubCmd = thisSubCmd.replace('log_out', str(logOutFile))
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/shower_simulator.py:236]: Replace job submission class in  PR #308

File: simtools/shower_simulator.py [hunk 6]
```diff
@@ -227,12 +227,48 @@ def submit(
...
+
+        meanRuntime = np.mean(runtime)
+
+        resources = dict()
+        resources['#events/run'] = nEvents
+        resources['Runtime/run [sec]'] = meanRuntime
+        resources['Runtime/1000 events [sec]'] = meanRuntime * 1000 / nEvents
+        return resources
+
+    def printResourcesReport(self):
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/shower_simulator.py:264]: Yes, it is very much the same and it should be generalized. Actually shower_simulator.py and array_simulator.py are almost identical and should be merged. I am opening an issue on that.

**Comment by GernotMaier** [comment]
@orelgueta - I think we are good for a second round of review.

File: applications/production.py [hunk 7]
```diff
@@ -55,10 +55,81 @@
 from simtools.array_simulator import ArraySimulator
 
 
-def proccessConfigFile(configFile, primaryConfig):
+def parse(description=None):
```
[the remark this answers is not in the document]
Comment (orelgueta) [reply applications/production.py:58]: This was missed

File: simtools/array_simulator.py [hunk 8]
```diff
@@ -262,29 +269,33 @@ def _makeInputList(inputFileList):
...
         Finds the run number for a given input file name.
-        Input file names must follow 'run1234_*' pattern.
+        Input file names should follow any pattern starting at
+        runNumberIndex followed by a '_' (e.g., 'run1234_*' pattern
+        using runNumberIndex=3 as default)
         If not found, returns 1.
         """
         fileName = str(Path(file).name)
-        runStr = fileName[3 : fileName.find("_")]
+        runStr = fileName[runNumberIndex : fileName.find("_", runNumberIndex)]
```
[the remark this answers is not in the document]
Comment (orelgueta) [reply simtools/array_simulator.py]: Looks so much nicer now =)

File: simtools/corsika/corsika_runner.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (orelgueta) [reply simtools/corsika/corsika_runner.py:373]: Really? I thought you go through CORSIKA and or sim_telarray log files which can be thousands of lines long.

Either way, we don't have to change to the mmap system, but I do suggest to at least search in reverse order as I assume that it's much quicker.

File: applications/production.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (GernotMaier) [reply applications/production.py:58]: I don't understand - it is there when looking at the source code (e.g. [here](https://github.com/gammasim/gammasim-tools/blob/c278b5a0cc69850e7a1d59c87d94fc8739a8a6a5/applications/production.py#L58) and at my version in the terminal, but it doesn't appear when looking at the pull request. Is this a github issue?

File: simtools/corsika/corsika_runner.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/corsika/corsika_runner.py:373]: This is whats happening:

- script submitted to farm has a `echo "RUNTIME: $SECONDS"` as last line, this prints out the wall time to stdout
- the stdout is captures by gridengine for me e.g., in a file in simtools-output/test-production/corsika/logs/log-sub-corsika-run6-Prod5-North-gamma_test-production-out.log:
```
RUNTIME: 16
```
So this one has a single line! If you re-run it and have another run with runnumber 6, you get a second line (this is why we look always at the last line of this file to get the last result).

File: applications/production.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (orelgueta) [reply applications/production.py:58]: OK, no matter, as long as it appears in the actual code (which now I saw in the branch).

File: simtools/corsika/corsika_runner.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (orelgueta) [reply simtools/corsika/corsika_runner.py:373]: OK, so this is definitely different to what I was expecting. This search in the file is perfectly fine then and this is discussion can be resolved.

**Pull request merged**